### PR TITLE
Add helpers for common setup code

### PR DESCRIPTION
**Pull request by cpuguy83**
- Create a clientset
- Setup pod informer filters
- Setup node lease client
- Add V2 naive node provider for accepting node status updates and use that to wait for node readiness in e2e tests

This splits up some stuff from #859

### CODE REVIEW

File: test/e2e/suite.go [hunk 1]
```diff
@@ -40,14 +43,29 @@ type EndToEndTestSuiteConfig struct {
...
 	if err := ts.setup(); err != nil {
-		panic(err)
+		t.Fatal(err)
 	}
 
-	// Wait for the virtual kubelet (deployed as a pod) to become fully ready
-	if _, err := f.WaitUntilPodReady(f.Namespace, f.NodeName); err != nil {
-		panic(err)
+	// Wait for the virtual kubelet node resource to become fully ready
+	if err := f.WaitUntilNodeCondition(func(ev watch.Event) (bool, error) {
```
Comment (sargun) [inline test/e2e/suite.go:52]: Isn't there literallly a Kubernetes wait function to do this?

File: cmd/virtual-kubelet/internal/commands/root/root.go [hunk 2]
```diff
@@ -228,33 +225,39 @@ func runRootCommand(ctx context.Context, s *provider.Store, c Opts) error {
 		}
 	}()
 
+	setNodeReady(pNode)
```
[the remark this answers is not in the document]
Comment (cpuguy83) [reply cmd/virtual-kubelet/internal/commands/root/root.go:228]: Are people using this code? I thought we only have this for e2e tests.

File: cmd/virtual-kubelet/internal/commands/root/root.go [hunk 3]
```diff
@@ -120,7 +116,7 @@ func runRootCommand(ctx context.Context, s *provider.Store, c Opts) error {
 		NodeName:          c.NodeName,
 		OperatingSystem:   c.OperatingSystem,
 		ResourceManager:   rm,
-		DaemonPort:        int32(c.ListenPort),
+		DaemonPort:        c.ListenPort,
```
[the remark this answers is not in the document]
Comment (cpuguy83) [reply cmd/virtual-kubelet/internal/commands/root/root.go:119]: It's already int32.

[the remark this answers is not in the document]
Comment (sargun) [reply cmd/virtual-kubelet/internal/commands/root/root.go:119]: oh, nevermind.

File: cmd/virtual-kubelet/internal/commands/root/root.go [hunk 4]
```diff
@@ -228,33 +225,39 @@ func runRootCommand(ctx context.Context, s *provider.Store, c Opts) error {
 		}
 	}()
 
+	setNodeReady(pNode)
+	if err := np.UpdateStatus(ctx, pNode); err != nil {
+		return errors.Wrap(err, "error marking the node as ready")
```
Comment (sargun) [inline cmd/virtual-kubelet/internal/commands/root/root.go:230]: 1.13 errors?

File: cmd/virtual-kubelet/internal/commands/root/root.go [hunk 5]
```diff
@@ -228,33 +225,39 @@ func runRootCommand(ctx context.Context, s *provider.Store, c Opts) error {
...
+		c.Reason = "KubeletReady"
+		c.Status = corev1.ConditionTrue
+		c.LastHeartbeatTime = metav1.Now()
+		c.LastTransitionTime = metav1.Now()
+		n.Status.Conditions[i] = c
+		return
 	}
 
-	return kubernetes.NewForConfig(config)
+	// No ready condition in node status
```
Comment (sargun) [inline cmd/virtual-kubelet/internal/commands/root/root.go:253]: We should add the default diskpressure, memorypressure, and (I think pid pressure) to this.

File: node/node.go [hunk 6]
```diff
@@ -648,7 +648,52 @@ func (NaiveNodeProvider) Ping(ctx context.Context) error {
...
 // function is a no-op.
-func (NaiveNodeProvider) NotifyNodeStatus(ctx context.Context, f func(*corev1.Node)) {
+func (n NaiveNodeProvider) NotifyNodeStatus(_ context.Context, _ func(*corev1.Node)) {
+}
+
+// NaiveNodeProviderV2 is like NaiveNodeProvider except it supports accepting node status updates.
+// It must be used with as a pointer and must be created with `NewNaiveNodeProvider`
+type NaiveNodeProviderV2 struct {
+	notify      func(*corev1.Node)
+	updateReady chan struct{}
```
[the remark this answers is not in the document]
Comment (sargun) [reply node/node.go:658]: Yeah, still think adding a comment here would be valuable.

File: node/nodeutil/client.go [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+func PodInformerFilter(node string) kubeinformers.SharedInformerOption {
+	return kubeinformers.WithTweakListOptions(func(options *metav1.ListOptions) {
+		options.FieldSelector = fields.OneTermEqualSelector("spec.nodeName", node).String()
+	})
+}
+
+// NodeLeaseV1Beta1Client creates a v1beta1 Lease client for use with node leases from the passed in client.
+//
+// Use this with node.WithNodeEnableLeaseV1Beta1 when creating a node controller.
+func NodeLeaseV1Beta1Client(client kubernetes.Interface) v1beta1.LeaseInterface {
```
Comment (sargun) [inline node/nodeutil/client.go:56]: It's likely for non-v1beta1, we'll have to pass the entire clientset.

File: test/e2e/suite.go (hunk 1 above)
Comment (cpuguy83) [reply test/e2e/suite.go:52]: Do you have a ref for this?

File: cmd/virtual-kubelet/internal/commands/root/root.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (cpuguy83) [reply cmd/virtual-kubelet/internal/commands/root/root.go:229]: @sargun are you using this code for anything other than e2e?

[the remark this answers is not in the document]
Comment (sargun) [reply cmd/virtual-kubelet/internal/commands/root/root.go:229]: Nope. We moved the code out of here.

File: test/e2e/suite.go (hunk 1 above)
Comment (sargun) [reply test/e2e/suite.go:52]: More, like, could we use the wait primitives? https://godoc.org/k8s.io/apimachinery/pkg/util/wait

Comment (cpuguy83) [reply test/e2e/suite.go:52]: @sargun So you don't want to use a watch here? Sorry I'm just confused as to what you are looking for specifically.

Comment (sargun) [reply test/e2e/suite.go:52]: Actually scratch that. For some reason, I thought the wait package had something where you could pass a conditionfunc to look at the results of a watch

File: cmd/virtual-kubelet/internal/commands/root/root.go (hunk 5 above)
Comment (cpuguy83) [reply cmd/virtual-kubelet/internal/commands/root/root.go:254]: I'd rather just take this out. If it fails its a provider problem (and specifically this is for e2e in this repo only anyway).

Comment (cpuguy83) [reply cmd/virtual-kubelet/internal/commands/root/root.go]: Updated, took this special condition out since it's not going to be reached anyway.

**Comment by sargun** [comment]
@cpuguy83 Are you going to merge this? I wanted to get https://github.com/virtual-kubelet/virtual-kubelet/pull/870/files in, and it's waiting on this.

**Pull request merged**